protocol-specs/utils/chain_state_builder.js

```diff
@@ -0,0 +1,282 @@
+					this.state.pendingTransactions.push(transferTx);
+					return this;
+				},
+			}),
+		};
+	}
+
+	registerDelegate(delegateName) {
+		return {
+			for: delegateAddress => {
+				const registerDelegateTx = new DelegateTransaction(
+					registerDelegate({
+						username: delegateName,
+						passphrase: Object.values(this.state.accounts).find(
+							anAccount => anAccount.address === delegateAddress,
+						).passphrase,
+					}),
+				);
+				// Push it to pending transaction
+				this.state.pendingTransactions.push(registerDelegateTx);
+				return this;
+			},
+		};
+	}
+
+	forge() {
+		const latestsAccountState = this.state.accountStore.slice(-1)[0];
+		this.processBlockTransactions(this.state.pendingTransactions);
+
+		const newBlock = createBlock(
+			defaultConfig,
+			latestsAccountState,
+			this.previousBlock,
+			this.round,
+			this.slot,
+			{
+				version: 1,
+				transactions: [...this.state.pendingTransactions],
+			},
+		);
+
+		this.state.chain.push(newBlock);
+		this.previousBlock = newBlock;
+		this.state.appliedTransactions.push([...this.state.pendingTransactions]);
+		this.state.pendingTransactions = [];
+		return this;
+	}
+
+	forgeInvalidInputBlock() {
+		const latestsAccountState = this.state.accountStore.slice(-1)[0];
+
+		const newBlock = createBlock(
+			defaultConfig,
+			latestsAccountState,
+			this.previousBlock,
+			this.round,
+			this.slot,
+			{
+				version: 1,
+				transactions: [...this.state.pendingTransactions],
+			},
+		);
+
+		this.state.inputBlock.push(newBlock);
+		this.state.appliedTransactions.push(...this.state.pendingTransactions);
+		this.state.pendingTransactions = [];
+		return this;
+	}
+
+	processBlockTransactions() {
+		// eslint-disable-next-line no-restricted-syntax
+		for (const aTransaction of this.state.pendingTransactions) {
+			switch (aTransaction.type) {
+				case 0:
+					this.updateAccountBalancesAfterTransfer(
+						aTransaction.senderId,
+						aTransaction.recipientId,
+						aTransaction.amount.toString(),
+					);
+					break;
+				case 2:
+					this.updateAccountStateAfterDelegateRegistration(
+						aTransaction.senderId,
+						this.fees.delegate,
+						aTransaction.asset.delegate.username,
+					);
+					break;
+				default:
+					break;
+			}
+		}
+		return this;
+	}
+
+	getScenario() {
+		return {
+			initialAccountsState: this.state.initialAccountStore,
+			finalAccountsState: this.state.accountStore,
+			chain: this.state.chain,
+			inputBlock: this.state.inputBlock,
+		};
+	}
+
+	updateAccountBalancesAfterTransfer(from, to, amount) {
+		const newAccountStoreState = cloneDeep(
+			this.state.accountStore.slice(-1)[0],
+		);
+		const sender = this.findAccountByAddress(from, newAccountStoreState);
+		const recipient = this.findAccountByAddress(to, newAccountStoreState);
+
+		if (!sender) {
+			throw new Error(
+				'Sender does not exists so it would not be possible to transfer form this account. Check the values passed to the constructor',
```

Comment:
```suggestion
				'Sender does not exists so it would not be possible to transfer from this account. Check the values passed to the constructor',
```